Backup: Copy backup-helper endpoints to backup package

fgiannar opened the pull request:
In order for the Backups plugin to fully work without Jetpack, we need to move the backup related endpoints to the `backup` package so that they can be consumed by both plugins independently.

#### Changes proposed in this Pull Request:
<!--- Explain what functional changes your PR includes -->
*  Backup Endpoints migrated to v4 REST API

#### Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->
1199658893173824-as-1200402244140846/f

#### Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->
No

#### Testing instructions:
- Create a JN site from this PR and make sure it's fully connected (site connected and admin user connected)
- Follow the instructions here: 29934-pb/#php

fgiannar commented on projects/packages/backup/src/class-rest-controller.php:
```diff
@@ -0,0 +1,141 @@
+		// Delete a Backup Helper Script.
+		register_rest_route(
+			'jetpack/v4',
+			'/delete-backup-helper-script',
```
I wanted a more RESTfull approach here, aka:
Install Helper Script: `POST /backup-helper-script`
Delete Helper Script: `DELETE /backup-helper-script`

However Jetpack Connection [restricts us](https://github.com/Automattic/jetpack/blob/master/projects/packages/connection/src/class-rest-authentication.php#L133) from using any HTTP methods other than `GET` or `POST`.

I also tried to call the `delete` endpoint using `POST` and `_method=DELETE` but that leaves us with 2 exactly same endpoint definitions. Am I missing something here? Any suggestion is more than welcome.

kbrown9 replied:
If you send a POST request with `_method=DELETE`, wouldn't you have two endpoints, one with `'methods' => WP_REST_Server::CREATABLE` and one with `'methods' => WP_REST_Server::DELETABLE`?

fgiannar replied:
Right! For some reason it was always failing when I was testing, with `rest_route_not_found`.  This made me wonder whether I should accept both `POST` and `DELETE` methods in the endpoint definition, which didn't sound right though.
I tried once again today using `X-HTTP-Method-Override` and worked so I modified the PR and the one bin script accordingly: 29958-pb/#php

fgiannar commented on projects/packages/backup/tests/php/test-rest-controller.php:
```diff
@@ -0,0 +1,246 @@
+		$this->server   = $wp_rest_server;
+
+		// Register REST routes.
+		add_action( 'rest_api_init', array( 'Automattic\\Jetpack\\Backup\\REST_Controller', 'register_rest_routes' ) );
```
I had to add this here because `actions.php` is auto-loaded by composer and when running unit tests at the time it gets loaded, WP is not available yet (exiting early because `ABSPATH` is not defined).

Now I wonder if this is ok in general or the way `actions.php` is loaded could cause any issues outside of unit testing.

cc @kbrown9  Just to confirm!

kbrown9 replied:
I can't think of any issues that would be caused by the way `actions.php` is loaded.

Are you concerned that the same issue that you're facing with unit testing could also happen outside of unit testing? I don't think that's possible. A plugin will load the autoloader, which will load the `actions.php` file. The plugin itself will be loaded by WP. I can't think of a situation where the `actions.php` file would get loaded before WP.

fgiannar replied:
That was my understanding as well, just wanted to make sure I'm not missing anything! Many thanks for confirming, Kim!

anomiex commented on projects/packages/backup/.gitattributes:
```diff
@@ -1,6 +1,9 @@
 
 # Files not needed in the production build.
-/changelog/**    production-exclude
+/changelog/**    production-exclude
```
Total nitpick, but would you mind adding the trailing newlines to these files? Look for that icon.

If you don't want to I won't insist, but some future PR touching the file will probably wind up adding it as an unrelated change.

fgiannar replied:
Sure! Thanks for pointing that out, Brad! Fixed.

samiff commented:
@fgiannar Would I be able to test this locally rather than on JN? I tried:

* Pulling down this branch to my local monorepo (using Jurassic Tube)
* JP is deactivated and site/user connection established via the Backup plugin
* In the Backup plugin requiring the `automattic/jetpack-backup` package
* The `register_rest_route` seem to be happening
* Though when I try to `install_backup_helper_script()` from sandbox with blog_token I get:

```
(object) array(
   'code' => 'rest_forbidden',
   'message' => 'You are not allowed to perform this action.',
   'data' =>
  (object) array(
     'status' => 401,
  ),
)
```

kbrown9 commented:
> @fgiannar Would I be able to test this locally rather than on JN? I tried:
> 
> * Pulling down this branch to my local monorepo (using Jurassic Tube)
> * JP is deactivated and site/user connection established via the Backup plugin
>   
>   * Fixing the Backup plugin conflicting package name [here](https://github.com/Automattic/jetpack/pull/20084/commits/a2ff83742fd547cc2613203939e585961e7c9b6c)
> * In the Backup plugin requiring the `automattic/jetpack-backup` package
> * The `register_rest_route` seem to be happening
> * Though when I try to `install_backup_helper_script()` from sandbox with blog_token I get:
> 
> ```
> (object) array(
>    'code' => 'rest_forbidden',
>    'message' => 'You are not allowed to perform this action.',
>    'data' =>
>   (object) array(
>      'status' => 401,
>   ),
> )
> ```

I tested this out, and the problem appears to be that we're not initializing the connection's `Rest_Authentication` hooks early enough in the Backup plugin.

I added this line to the `Jetpack_Backup` class:
`use Automattic\Jetpack\Connection\Rest_Authentication as Connection_Rest_Authentication;`

and called `Connection_Rest_Authentication::init();` in the `Jetpack_Backup` constructor. That seemed to resolve the authentication problem. I'm not sure if this is how we'll want to initialize the rest authentication hooks, but it allowed me to test the endpoints with the Backup plugin alone.

fgiannar commented:
@samiff @kbrown9 Many thanks for testing this with the Backups Plugin! Also, Kim, great job figuring out how to make those endpoints work with the Backup plugin alone!
However, this is out of the current PR's scope and I was planning to address it as a separate issue if you agree.

jeherve reviewed (commented):
This worked well before your latest change, but I can't seem to be able to delete the helper scripts after the change. I updated the call to `$this->call_core_api( 'DELETE', '/jetpack/v4/backup-helper-script' );`, but that hasn't helped. Am I missing something?

fgiannar commented:
> This worked well before your latest change, but I can't seem to be able to delete the helper scripts after the change. I updated the call to `$this->call_core_api( 'DELETE', '/jetpack/v4/backup-helper-script' );`, but that hasn't helped. Am I missing something?

Thanks for reviewing, Jeremy! I've also updated the one bin script [accordingly](https://github.com/Automattic/jetpack/pull/20097#discussion_r654285063). Did you use the updated version here 29958-pb/#php?

jeherve reviewed (approved):
This seems to work well for me now. 🚢
